src/core/ad.jl

```diff
@@ -1,127 +1,81 @@
 """
-    gradient(vi::VarInfo, model::Function, spl::Union{Nothing, Sampler})
-
-Function to generate the gradient dictionary, with each prior map to its derivative of the logjoint probibilioty. This function uses chunk-wise forward AD with a chunk of size $(CHUNKSIZE) as default.
-
-Example:
-
-```julia
-grad = gradient(vi, model, spl)
-end
-```
+gradient_forward(
+    θ::AbstractVector{<:Real},
+    vi::VarInfo,
+    spl::Sampler,
+    model::Function,
+)
+
+Computes the gradient of the log joint of `θ` for the model specified by `(vi, spl, model)`
+using forwards-mode AD from ForwardDiff.jl.
 """
-gradient(vi::VarInfo, model::Function) = gradient(vi, model, nothing)
-gradient(vi::VarInfo, model::Function, spl::Union{Nothing, Sampler}) = begin
-
-  θ_hash = hash(vi[spl])
-
-  if spl != nothing && haskey(spl.info, :grad_cache)
-    if haskey(spl.info[:grad_cache], θ_hash)
-      return spl.info[:grad_cache][θ_hash]
-    end
+function gradient_forward(
+  θ::AbstractVector{<:Real},
+  vi::VarInfo,
+  model::Function,
+  spl::Union{Nothing, Sampler}=nothing,
+)
+  # Record old parameters.
+  θ_old = vi[spl]
+
+  # Define function to compute log joint.
+  function f(θ)
+      vi[spl] = θ
+      return -runmodel(model, vi, spl).logp
   end
 
-  # Initialisation
-  grad = Vector{Float64}()
+  # Set chunk size and do ForwardMode.
+  chunk = ForwardDiff.Chunk(min(CHUNKSIZE, length(θ)))
+  config = ForwardDiff.GradientConfig(f, θ, chunk)
+  ∂l∂θ = ForwardDiff.gradient!(similar(θ), f, θ, config)
 
-  # Split keys(vi) into chunks,
-  @debug "making chunks..."
-  vn_chunk = Set{VarName}(); vn_chunks = []; chunk_dim = 0;
+  # Replace old parameters to ensure this function doesn't mutate `vi`.
+  vi[spl] = θ_old
 
-  vns = getvns(vi, spl); vn_num = length(vns)
+  return ∂l∂θ
```

Comment:
I think we should make the return between forward and reverse mode consistent.

Reply:
By which I mean `return getlogp(vi), ∂l∂θ` here

Reply:
I entirely agree -- see TODOs above.